akka-cluster-sharding/src/main/scala/akka/cluster/sharding/SliceRangeShardAllocationStrategy.scala

```diff
@@ -0,0 +1,264 @@
+/*
+ * Copyright (C) 2024-2023 Lightbend Inc. <https://www.lightbend.com>
+ */
+
+package akka.cluster.sharding
+
+import scala.annotation.tailrec
+import scala.collection.immutable
+import scala.concurrent.Future
+
+import akka.actor.ActorRef
+import akka.actor.ActorSystem
+import akka.cluster.Cluster
+import akka.cluster.ClusterEvent.CurrentClusterState
+import akka.cluster.Member
+import akka.cluster.sharding.ShardCoordinator.ActorSystemDependentAllocationStrategy
+import akka.cluster.sharding.ShardRegion.ShardId
+import akka.cluster.sharding.internal.ClusterShardAllocationMixin
+import akka.cluster.sharding.internal.ClusterShardAllocationMixin.RegionEntry
+import akka.cluster.sharding.internal.ClusterShardAllocationMixin.ShardSuitabilityOrdering
+
+object SliceRangeShardAllocationStrategy {
+  // Not using Persistence because akka-persistence dependency could be optional
+  private val NumberOfSlices = 1024
+
+  private val emptyRebalanceResult = Future.successful(Set.empty[ShardId])
+
+  final class ShardBySliceMessageExtractor // FIXME
+}
+
+/**
+ * It will not rebalance when there is already an ongoing rebalance in progress.
+ *
+ * @param absoluteLimit the maximum number of shards that will be rebalanced in one rebalance round
+ * @param relativeLimit fraction (< 1.0) of total number of (known) shards that will be rebalanced
+ *                      in one rebalance round
+ */
+class SliceRangeShardAllocationStrategy(absoluteLimit: Int, relativeLimit: Double)
+    extends ActorSystemDependentAllocationStrategy
+    with ClusterShardAllocationMixin {
+  import SliceRangeShardAllocationStrategy._
+
+  private var cluster: Cluster = _
+
+  override def start(system: ActorSystem): Unit = {
+    cluster = Cluster(system)
+  }
+
+  override protected def clusterState: CurrentClusterState = cluster.state
+  override protected def selfMember: Member = cluster.selfMember
+
+  private var previousRebalance = Set.empty[ShardId]
+
+  override def allocateShard(
+      requester: ActorRef,
+      shardId: ShardId,
+      currentShardAllocations: Map[ActorRef, IndexedSeq[ShardId]]): Future[ActorRef] = {
+    val slice = shardId.toInt
+    if (slice >= NumberOfSlices)
+      throw new IllegalArgumentException("slice must be between 0 and 1023. Use `ShardBySliceMessageExtractor`.")
+
+    // allow some overfill, rebalance will try to move from the lower/upper slices
+    val overfill = 2
+    val maxShards = (NumberOfSlices / currentShardAllocations.size) + overfill
+
+    // FIXME take a look at ShardSuitabilityOrdering for member status and appVersion preference
```

Comment:
This seems quite important, or I think it will try to stick to allocating to old nodes until quite a number of cluster nodes has rolled, as it prefers neighbours. Maybe I'm missing something with the maxShards protecting against that?